migration/src/main/java/kin/sdk/migration/MigrationManager.java

```diff
@@ -1,60 +1,404 @@
+			@Override
+			protected String getIssuerAccountId() {
+				return migrationNetworkInfo.getIssuer();
+			}
+		};
+		return new KinClientCoreImpl(context, environment, appId, storeKey);
+	}
+
+	@NonNull
+	private IKinClient initNewKin() {
+		IKinClient kinClient;
+		Environment environment = new Environment(migrationNetworkInfo.getSdkNetworkUrl(),
+			migrationNetworkInfo.getSdkNetworkId());
+		kinClient = new KinClientSdkImpl(context, environment, appId, storeKey);
+		return kinClient;
+	}
+
+	private void saveMigrationCompleted() {
+		// save migration completion status from the persistent state.
+		SharedPreferences sharedPreferences = getSharedPreferences();
+		sharedPreferences.edit().putBoolean(MIGRATION_COMPLETED_KEY, true).apply();
+	}
+
+	private boolean isMigrationAlreadyCompleted() {
+		// get migration completion status from the persistent state.
+		SharedPreferences sharedPreferences = getSharedPreferences();
+		return sharedPreferences.getBoolean(MIGRATION_COMPLETED_KEY, false);
+	}
+
+	private SharedPreferences getSharedPreferences() {
+		return context.getSharedPreferences(MIGRATION_MODULE_PREFERENCE_FILE_KEY, Context.MODE_PRIVATE);
+	}
+
+	private void sendRequest(String url, Callback callback) {
+		Request request = new Request.Builder()
+			.url(url)
+			.post(RequestBody.create(null, ""))
+			.build();
+		okHttpClient.newCall(request).enqueue(callback);
+	}
+
+	private void fireOnError(final MigrationManagerListener migrationManagerListener, final Exception e) {
+		Log.d(TAG, "fireOnError: ");
+		handler.post(new Runnable() {
+			@Override
+			public void run() {
+				if (migrationManagerListener != null) {
+					inMigrationProcess.set(false);
+					migrationManagerListener.onError(e);
+				}
+			}
+		});
+	}
+
+	private void fireOnReady(final MigrationManagerListener migrationManagerListener, final IKinClient kinClient,
+		final boolean needToSave) {
+		Log.d(TAG, "fireOnReady: ");
+		handler.post(new Runnable() {
+			@Override
+			public void run() {
+				if (migrationManagerListener != null) {
+					if (needToSave) {
+						saveMigrationCompleted();
+					}
+					inMigrationProcess.set(false);
+					migrationManagerListener.onReady(kinClient);
+				}
+			}
+		});
+	}
+
+	private static class RetryInterceptor implements Interceptor {
+
+		private int retryCounter = 1;
+
+		@Override
+		public Response intercept(@NonNull Chain chain) throws IOException {
+			Log.d(TAG, "intercept: ");
+			Request request = chain.request();
+			Response response = chain.proceed(request);
+			if (response.code() >= 500) {
+				return retryRequest(request, chain);
+			}
+			Log.d(TAG, "INTERCEPTED: " + response.toString());
+			return response;
+		}
+
+		private Response retryRequest(Request req, Chain chain) throws IOException {
+			Log.d(TAG, "Retrying new request");
+			Request newRequest = req.newBuilder().build();
+			Response another = chain.proceed(newRequest);
+			while (retryCounter < MAX_RETRIES && another.code() >= 500) {
```

Comment:
The field is only shared in the interceptor class and is getting set to 1 when finished.
Although after you mentioned it i start testing(wanted to test it only after Tuesday) and it seems that i need to set it to one only after finished with the intercept method and not in the retry itself.
Regarding the retry itself:
The method is recursive because it was easy for me to think recursively in this case
Second, only if the response was 500 and bigger than a retry will happen.
First there will be an immediate call(a retry) and then starting a loop of retries.
For example if we want to have 3 retries then the first one will occur and then w while loop which will cause to another max of 2 retries.
If you still don't understand or think there is an error then let us do a review for this method next to each other.

Reply:
The recursion is the least problem here, I don't mind keeping it. 
I still think the retry should be var in the method for safety, and still, you will return only first response and ignore the retry response, which means retry will be worthless.

Reply:
We will talked about it when we will meet.